src/Domain/DiagnosticInfo.hpp

```diff
@@ -0,0 +1,109 @@
+  ASSERT(total_number_of_grid_points ==
+             alg::accumulate(grid_points_per_core, 0_st),
+         "Number of grid points determined from grid_points_per_core ("
+             << alg::accumulate(grid_points_per_core, 0_st)
+             << ") does not match number of grid points determined from "
+                "grid points_per_node ("
+             << total_number_of_grid_points << ").");
+  ASSERT(number_of_cores == elements_per_core.size(),
+         "Number of cores determined from elements_per_core ("
+             << elements_per_core.size()
+             << ") does not match number of cores determined from "
+                "global cache ("
+             << number_of_cores << ").");
+  ASSERT(number_of_cores == grid_points_per_core.size(),
+         "Number of cores determined from grid_points_per_core ("
+             << grid_points_per_core.size()
+             << ") does not match number of cores determined from "
+                "global cache ("
+             << number_of_cores << ").");
+  ASSERT(number_of_nodes == elements_per_node.size(),
+         "Number of nodes determined from elements_per_node ("
+             << elements_per_node.size()
+             << ") does not match number of nodes determined from "
+                "global cache ("
+             << number_of_nodes << ").");
+  ASSERT(number_of_nodes == grid_points_per_node.size(),
+         "Number of nodes determined from grid_points_per_node ("
+             << grid_points_per_node.size()
+             << ") does not match number of nodes determined from "
+                "global cache ("
+             << number_of_nodes << ").");
+
+  using ::operator<<;
+  std::stringstream ss{};
+  ss << "----- Domain Info -----\n";
+  ss << "Total blocks: " << total_number_of_blocks << "\n";
+  ss << "Total elements: " << total_number_of_elements << "\n";
+  ss << "Total grid points: " << total_number_of_grid_points << "\n";
+  ss << "Number of cores: " << number_of_cores << "\n";
+  ss << "Number of nodes: " << number_of_nodes << "\n";
+  ss << "Elements per core: " << elements_per_core << "\n";
+  ss << "Elements per node: " << elements_per_node << "\n";
+  ss << "Grid points per core: " << grid_points_per_core << "\n";
+  ss << "Grid points per node: " << grid_points_per_node << "\n";
+  ss << "-----------------------\n";
+
+  return ss.str();
+}
+
+/// Prints domain diagnostic information using the `diagnostic_info()` function.
+template <size_t Dim, typename Metavariables>
+void print_diagnostic_info(const Domain<Dim>& domain,
+                           const Parallel::GlobalCache<Metavariables>& cache,
+                           const std::vector<size_t>& elements_per_core,
+                           const std::vector<size_t>& elements_per_node,
+                           const std::vector<size_t>& grid_points_per_core,
+                           const std::vector<size_t>& grid_points_per_node) {
+  Parallel::printf(
```

Comment:
I generally don't like things that print, especially something like this that just prints a string. It's generally better to just have the end user decide where to place the string. E.g. stream operators don't do any `std::cout`-type stuff internally. The change request is: just remove `print_diagnostic_info` and call `Parallel::printf(...)` wherever you need to print.